Problem with setting up the project - undefined reference to symbol 'XFree'

Hi,

i'm kinda sorry to open this as issue, because i think the fault lies by with me - not in this project. But i could not think of another place where i could find help to this

i'm using ubuntu 17.04, and CLion as my IDE.

i created a new C++ Project. it contains my main.cpp, a CMakeLists.txt and a subfolder thirdparty.

i clone the bgfx.cmake repository into the thirdparty subfolder and changed my CMakeLists.txt to look like this:

```
cmake_minimum_required(VERSION 3.7)
set(CMAKE_CXX_STANDARD 14)
project(bgfx_test)

set(SOURCE_FILES main.cpp)

add_subdirectory("thirdparty/bgfx.cmake")
add_executable(bgfx_test ${SOURCE_FILES} )

target_link_libraries(bgfx_test bx)
target_link_libraries(bgfx_test bgfx)
target_link_libraries(bgfx_test bimg)

#target_include_directories(bgfx_test PUBLIC "cmake-build-debug/thirdparty/bgfx.cmake")
```

Then i changed my main.cpp to look like the HelloWorld example but with an empty update loop:
```
#include <bx/uint32_t.h>
#include <bgfx/bgfx.h>
#include "thirdparty/bgfx.cmake/bgfx/examples/common/common.h"
#include "thirdparty/bgfx.cmake/bgfx/examples/common/bgfx_utils.h"

class ExampleHelloWorld : public entry::AppI {
    int shutdown() BX_OVERRIDE {
        return 0;
    }

    bool update() BX_OVERRIDE {
        return false;
    }

    uint32_t m_width;
    uint32_t m_height;
    uint32_t m_debug;
    uint32_t m_reset;

    void init(int _argc, char** _argv) BX_OVERRIDE
    {
        Args args(_argc, _argv);

        m_width  = 1280;
        m_height = 720;
        m_debug  = BGFX_DEBUG_TEXT;
        m_reset  = BGFX_RESET_VSYNC;

        bgfx::init(args.m_type, args.m_pciId);
        bgfx::reset(m_width, m_height, m_reset);

        // Enable debug text.
        bgfx::setDebug(m_debug);

        // Set view 0 clear state.
        bgfx::setViewClear(0
                , BGFX_CLEAR_COLOR|BGFX_CLEAR_DEPTH
                , 0x303030ff
                , 1.0f
                , 0
        );
    }
};

ENTRY_IMPLEMENT_MAIN(ExampleHelloWorld);
```

when i try to run the project i get this error:

> /usr/bin/ld: thirdparty/bgfx.cmake/libbgfxd.a(glcontext_glx.cpp.o): undefined reference to symbol 'XFree'
>
> //usr/lib/x86_64-linux-gnu/libX11.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status


as i wrote: i'm pretty sure i did something wrong, but i rly can't figure out what.
So maybe someone can push me into the right direction.

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,10 +2,18 @@ sudo: required
 dist: precise
 language: cpp
 
+os:
+  - osx
+
+env:
+  - AMALGAMATED=ON
+  - AMALGAMATED=OFF
+
 matrix:
   include:
-    - compiler: gcc
-      addons:
+    - os: linux
+      compiler: gcc
+      addons: &gcc-addons
         apt:
           sources:
             - ubuntu-toolchain-r-test
@@ -14,27 +22,30 @@ matrix:
             - g++-5
             - cmake
             - cmake-data
-      env: COMPILER=g++-5
-    - compiler: clang
-      addons:
+      env: COMPILER=g++-5 AMALGAMATED=ON
+    - os: linux
+      compiler: clang
+      addons: &clang-addons
         apt:
           sources:
             - ubuntu-toolchain-r-test
             - llvm-toolchain-precise-3.8
-            - george-edison55-precise-backports 
+            - george-edison55-precise-backports
           packages:
             - clang-3.8
             - cmake
             - cmake-data
-      env: COMPILER=clang++-3.8
-
-before_install:
-  - sudo apt-get update -qq
-  - sudo apt-get install -y libglew-dev
+      env: COMPILER=clang++-3.8 AMALGAMATED=ON
+    - os: linux
+      compiler: gcc
+      addons: *gcc-addons
+      env: COMPILER=g++-5 AMALGAMATED=OFF
+    - os: linux
+      compiler: clang
+      addons: *clang-addons
+      env: COMPILER=clang++-3.8 AMALGAMATED=OFF
 
 script:
-  - git submodule init
-  - git submodule update
-  - mkdir build
-  - cd build
-  - cmake -DCMAKE_INSTALL_PREFIX=./install -DCMAKE_CXX_COMPILER=$COMPILER .. && make
+  - mkdir build && cd build
+  - cmake -DCMAKE_CXX_COMPILER=$COMPILER -DBGFX_INSTALL_EXAMPLES=ON -DBGFX_AMALGAMATED=$AMALGAMATED -DBX_AMALGAMATED=$AMALGAMATED ..
+  - cmake --build .

--- a/cmake/bgfx.cmake
+++ b/cmake/bgfx.cmake
@@ -14,24 +14,28 @@ if( NOT IS_DIRECTORY ${BGFX_DIR} )
 	return()
 endif()
 
+if(NOT APPLE)
+	set(BGFX_AMALGAMATED_SOURCE ${BGFX_DIR}/src/amalgamated.cpp)
+else()
+	set(BGFX_AMALGAMATED_SOURCE ${BGFX_DIR}/src/amalgamated.mm)
+endif()
+
 # Grab the bgfx source files
 file( GLOB BGFX_SOURCES ${BGFX_DIR}/src/*.cpp ${BGFX_DIR}/src/*.mm ${BGFX_DIR}/src/*.h ${BGFX_DIR}/include/bgfx/*.h ${BGFX_DIR}/include/bgfx/c99/*.h )
 if(BGFX_AMALGAMATED)
 	set(BGFX_NOBUILD ${BGFX_SOURCES})
-	list(REMOVE_ITEM BGFX_NOBUILD ${BGFX_DIR}/src/amalgamated.cpp)
+	list(REMOVE_ITEM BGFX_NOBUILD ${BGFX_AMALGAMATED_SOURCE})
 	foreach(BGFX_SRC ${BGFX_NOBUILD})
 		set_source_files_properties( ${BGFX_SRC} PROPERTIES HEADER_FILE_ONLY ON )
 	endforeach()
 else()
-	if(APPLE)
-		set_source_files_properties( ${BGFX_DIR}/src/amalgamated.mm PROPERTIES HEADER_FILE_ONLY ON )
-	else()
-		set_source_files_properties( ${BGFX_DIR}/src/amalgamated.cpp PROPERTIES HEADER_FILE_ONLY ON )
-	endif()
+	# Do not build using amalgamated sources
+	set_source_files_properties( ${BGFX_DIR}/src/amalgamated.cpp PROPERTIES HEADER_FILE_ONLY ON )
+	set_source_files_properties( ${BGFX_DIR}/src/amalgamated.mm PROPERTIES HEADER_FILE_ONLY ON )
 endif()
 
 # Create the bgfx target
-add_library( bgfx STATIC ${BGFX_SOURCES} )
+add_library( bgfx ${BGFX_SOURCES} )
 
 # Enable BGFX_CONFIG_DEBUG in Debug configuration
 target_compile_definitions( bgfx PRIVATE "$<$<CONFIG:Debug>:BGFX_CONFIG_DEBUG=1>" )
@@ -68,11 +72,14 @@ if( APPLE )
 endif()
 
 if( UNIX AND NOT APPLE )
-	target_link_libraries( bgfx PUBLIC GL )
+	find_package(X11 REQUIRED)
+	find_package(OpenGL REQUIRED)
+	#The following commented libraries are linked by bx
+	#find_package(Threads REQUIRED)
+	#find_library(LIBRT_LIBRARIES rt)
+	#find_library(LIBDL_LIBRARIES dl)
+	target_link_libraries( bgfx PUBLIC ${X11_LIBRARIES} ${OPENGL_LIBRARIES})
 endif()
-
-# Excluded files from compilation
-set_source_files_properties( ${BGFX_DIR}/src/amalgamated.mm PROPERTIES HEADER_FILE_ONLY ON )
 
 # Exclude mm files if not on OS X
 if( NOT APPLE )

--- a/cmake/bx.cmake
+++ b/cmake/bx.cmake
@@ -65,6 +65,10 @@ if( UNIX AND NOT APPLE )
 
 	# Real time (for clock_gettime)
 	target_link_libraries( bx rt )
+elseif(APPLE)
+	find_library( FOUNDATION_LIBRARY Foundation)
+	mark_as_advanced( FOUNDATION_LIBRARY )
+	target_link_libraries( bx PUBLIC ${FOUNDATION_LIBRARY} )
 endif()
 
 # Put in a "bgfx" folder in Visual Studio

